Fix releasing a shared lock multiple times

jvsalo opened the pull request:
This patch makes `MemcacheLockingProvider` handle multiple releases of the same shared lock gracefully.

Previously, the lock will get negative values that cause problems as seen in https://github.com/nextcloud/server/issues/21073

During PUT requests, https://github.com/nextcloud/server/blob/5e35594cb6a8eeb89197ae2ac044bc352bbaba38/apps/dav/lib/Connector/Sabre/Directory.php#L157 releases a lock once, and https://github.com/nextcloud/server/blob/5e35594cb6a8eeb89197ae2ac044bc352bbaba38/apps/dav/lib/Connector/Sabre/LockPlugin.php#L83 releases the same lock again (see the issue); `LockPlugin` never acquired the lock.

`MemcacheLockingProvider` already had some logic to handle multiple releases (negative going lock values), but it was subject to a race condition.

jvsalo committed:
Fix releasing a shared lock multiple times

Signed-off-by: Jaakko Salo <jaakkos@gmail.com>

kesselb commented:
Thanks for your patch :tada: 

If the LockPlugin already obtains the lock why not kill the shared lock in createFile? :thinking:

jvsalo commented:
> If the LockPlugin already obtains the lock why not remove the shared lock in createFile?

Thanks for the quick response! In my debug session, I found that `LockPlugin::getLock()` doesn't get a `Node` from `getNodeForPath()`, and as such cannot call `acquireLock()`. However, upon exiting the request (`afterMethod` handler), `LockPlugin::releaseLock()` *does* get the `Node` and will indeed call `releaseLock()`.

I'm very new to the codebase so I only have little idea what I'm talking about, but this I discovered by instrumenting the code yesterday.

kesselb commented:
> Since it's a PUT request for a new file, perhaps it makes some sense that initially the Node is not found?

That makes sense. So it's maybe enough to remove the lock release from createFile?

jvsalo commented:
> So it's maybe enough to remove the lock release from createFile?

In the current code I like that it's clear what part is responsible for releasing each lock - it's literally all in the same file/function.

If `Directory::createFile()` lock is released in `LockPlugin::releaseLock()`, it may be surprising to someone studying/extending the locking code?

I find it a bit confusing that `LockPlugin` *maybe* acquires some lock and *maybe* releases it. Perhaps `LockPlugin` should remember if it actually got the lock, and only try to release if it actually did get it? AFAIK this should avoid double-locking completely in the "PUT new file" case. I suppose for updating a file, double-lock would still be taken.

jvsalo committed:
In LockPlugin, only release a lock if it was acquired

When uploading new files, getNodeForPath() will not succeed
yet so the lock cannot be acquired.

In that case, don't try to unlock it either.

Signed-off-by: Jaakko Salo <jaakkos@gmail.com>

MorrisJobke reviewed (approved):
Code looks good 👍

MorrisJobke commented:
/backport to stable19

MorrisJobke commented:
/backport to stable18

MorrisJobke commented:
/backport to stable17

MorrisJobke commented:
I closed the ones to 17 and 18, because they do not apply clean.
